### ISSUE TITLE
Public key pinning with SSLMate

### ISSUE THREAD

**Issue by konklone** [SSLMate/sslmate#10]
I'm evaluating [Public Key Pinning](https://developer.mozilla.org/en-US/docs/Web/Security/Public_Key_Pinning) for my personal and work websites. It's a [burgeoning web standard](https://tools.ietf.org/html/draft-ietf-websec-key-pinning-21) where the website can declare that it's pinned to some CA, or intermediate CA.

Have you thought about PKP in the context of SSLMate at all? I ask because SSLMate currently has two swappable backends, and presumably part of the value of SSLMate is the ability to add others behind the scenes. This is a bit in tension with the goals and guarantees of public key pinning.

Presumably, we should be targeting a root or intermediate, and not a leaf. One can pin to multiple certs. But, if we set a long max-age and pin to the two possible roots for SSLMate right now, then if SSLMate moves to a different CA in the future, we'd be stuck at keeping up until our key pinning max-age ran out.

What are your thoughts and recommendations on key pinning for SSLMate users? 

cc @seanherron @noahkunin

**Comment by AGWA** [comment]
Complicated question, without a good answer.  The short answer is that I think SSLMate can commit to:

1. Always reissuing a cert from the same CA.

2. Telling you, when the auto-renewal cycle starts 28 days before expiration, which CA will sign the renewed certificate.

Longer answer:

I would not recommend pinning an intermediate CA.  They seem too much like an implementation detail of CAs.  They changed during the SHA-2 transition, and I'd be worried about them changing again, even during a reissue.

I have reservations about pinning a root CA, because it makes it harder to change CAs.  That's not just a concern with SSLMate's use of multiple CAs, but a concern for website operators in general, who should have the freedom to take their business elsewhere.

Also, pinning a CA cert doesn't protect you against a compromise of your CA.  Pinning a root is especially weak, since roots often cross-sign other CAs, or sign an intermediate CA that has been sold to a government or a company that might not have the best security practices.

It's most secure to pin your own key.  To allow for rotation, it's essential to also pin several backup keys that you keep offline.  But that limits how frequently you can rotate, and would put you in a very bad situation if your backup keys all get compromised.

Personally, I'm disappointed that [TACK](http://tack.io/) didn't go anywhere.  Instead of pinning TLS public keys themselves, TACK lets you pin a server-chosen signing key that signs the TLS keys (kind of like being your own CA).  TACK would have let you pin your own key without the problems I described above.

That said, one really nice thing about HPKP is that it lets you specify a URL where pinning violations should be reported.  I think this provides a strong disincentive to attackers, who would prefer their attack go undetected, and an incentive to CAs to have good security practices.  So even if you have to use a shortish max-age to work around the problems described above, there's still a lot of value in deploying HPKP.  (This is assuming that browsers are robust about retrying failed report submissions, since an active attacker is probably also in a position to block submissions.)

SSLMate might be able to provide some additional services to help use HPKP effectively, such as by hosting an endpoint for violation reports, or by providing tools that manage your offline backup keys and generate the correct HPKP header to send.  I'll be thinking more about this, and would definitely appreciate input.  For instance, what kind of max-age were you hoping to use, and is pinning your own keys, rather than a CA's, something you'd consider?

**Comment by noncombatant** [comment]
Several things.

1. HPKP is too dangerous to do by default. Site operators should do it by hand, knowing for sure that they are doing it, and being certain that they mean to.

2. Just because the intermediate *cert* changes, doesn't mean the *key* changes. A pin is a function of the key, not of the cert. So those SHA-1 intermediates that were replaced by SHA-2 intermediates might just be the same key wrapped in a fancy new cert, and pins to such keys would not break.

3. You can pin to as many keys, at any place in the chain, as you want. Go ahead and pin the root(s), intermediate(s), and end-entity keys. The more keys you pin to, the better the site operator is defending against bricking their site (but now they potentially trust more issuers to issue for them, if they pinned many issuers).

In case of CA change, just have the new CA sign your existing EE key, and you're good to go (for example).

You can in fact get something similar to TACK by pinning EE keys (and backup EE keys — the spec requires backup keys anyway!).

**Comment by AGWA** [comment]
Thanks for chiming in @noncombatant.

> HPKP is too dangerous to do by default. Site operators should do it by hand, knowing for sure that they are doing it, and being certain that they mean to.

Agreed.  There is definitely a very real risk of bricking your site.  Hopefully with good tools the risk of human error can be diminished, but it's definitely not something to do by default.

> Just because the intermediate cert changes, doesn't mean the key changes. A pin is a function of the key, not of the cert. So those SHA-1 intermediates that were replaced by SHA-2 intermediates might just be the same key wrapped in a fancy new cert, and pins to such keys would not break.

In a sane world, keys would not change when they didn't have to.  But the intermediate certificate keys changed for both RapidSSL and PositiveSSL during the SHA-2 transition.

> You can pin to as many keys, at any place in the chain, as you want. Go ahead and pin the root(s), intermediate(s), and end-entity keys

That's an excellent idea for surviving a CA change.

**Comment by sleevi** [comment]
To chime in with my fellow HPKP-draft author, I did want to nit pick on one particular piece of advice (well, two really) from above:

> Pinning a root is especially weak, since roots often cross-sign other CAs, or sign an intermediate CA that has been sold to a government or a company that might not have the best security practices.
and
> It's most secure to pin your own key. To allow for rotation, it's essential to also pin several backup keys that you keep offline.

Though both of these are strictly true statements, taken together, they might lead people to believe that pinning the EE key is the best solution, when it's arguably the most dangerous (operationally), and that's the real risk (as @noncombatant notes)

We tried to capture this in the spec in https://tools.ietf.org/html/draft-ietf-websec-key-pinning-21#section-4 , and hopefully that says it better than I will here.

My recommendations for those considering deploying pinning are:
- Pin at the root first. You gain significant operational security benefits without as much of the attendant risk of bricking. You don't even need to exclusively pin to those you have business relationships with - you might choose to pin to several CAs that you might consider. You still greatly reduce your risk from third-party CAs (that aren't cross-signed), while still maintaining flexibility to go to other cert providers.
- Apply pressure through customer support channels to your CAs to encourage them to better support pinning their intermediates. It is a Mozilla program requirement that every CA document and disclose all of the intermediate certificates they have signed, and so hopefully within the next year as this information is made widely available, it will make it easier to make more informed pin sets.
- As noted by @noncombatant , pin widely.
